Do not hide compile-time thread safety warnings

hebasto opened the pull request:
On the way of transit from `RecursiveMutex` to `Mutex` (see #19303) it is crucial to have run-time `AssertLockHeld()` assertion that does _not_ hide compile-time Clang Thread Safety Analysis warnings.

On master (65e4ecabd5b4252154640c7bac38c92a3f3a7018) using `AssertLockHeld()` could hide Clang Thread Safety Analysis warnings, e.g., with the following patch applied:
```diff
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -607,7 +607,7 @@ public:
     void addUnchecked(const CTxMemPoolEntry& entry, setEntries& setAncestors, bool validFeeEstimate = true) EXCLUSIVE_LOCKS_REQUIRED(cs, cs_main);
 
     void removeRecursive(const CTransaction& tx, MemPoolRemovalReason reason) EXCLUSIVE_LOCKS_REQUIRED(cs);
-    void removeForReorg(const CCoinsViewCache* pcoins, unsigned int nMemPoolHeight, int flags) EXCLUSIVE_LOCKS_REQUIRED(cs, cs_main);
+    void removeForReorg(const CCoinsViewCache* pcoins, unsigned int nMemPoolHeight, int flags) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
     void removeConflicts(const CTransaction& tx) EXCLUSIVE_LOCKS_REQUIRED(cs);
     void removeForBlock(const std::vector<CTransactionRef>& vtx, unsigned int nBlockHeight) EXCLUSIVE_LOCKS_REQUIRED(cs);
 
```
Clang compiles the code without any thread safety warnings.

See "Add missed thread safety annotations" commit for the actual thread safety warnings that are fixed in this PR.

hebasto committed:
Use LockAssertion utility class instead of AssertLockHeld()

This change prepares for upcoming commit "Do not hide compile-time
thread safety warnings" by replacing AssertLockHeld() with
LockAssertion() where needed.

hebasto committed:
Do not hide compile-time thread safety warnings

hebasto committed:
Add missed thread safety annotations

This is needed for upcoming commit "sync.h: Make runtime lock checks
require compile-time lock checks" to pass.

hebasto commented:
Friendly ping @ajtowns @vasild

maflcko reviewed (commented):
weak Concept ACK. Will review some more later

hebasto commented:
Updated fa9157916514903cd7b526375597b1de0d8df8f0 -> 3f00515e92e7b6aed97c126592ad94b8a52a0e18 ([pr19668.01](https://github.com/hebasto/bitcoin/commits/pr19668.01) -> [pr19668.02](https://github.com/hebasto/bitcoin/commits/pr19668.02), [compare](https://github.com/hebasto/bitcoin/compare/pr19668.01..pr19668.02)).

This is an alternative approach without issues [mentioned](https://github.com/bitcoin/bitcoin/pull/19668#discussion_r465738894) by @MarcoFalke:
> It seems a layer-violation to import the validation lock cs_main into the net.h header to annotate this function. Also, this approach doesn't seem like it scales, since it has to be done for each mutex that is held during a call to `ForEachNode`.

hebasto commented:
@MarcoFalke 
> weak Concept ACK. Will review some more later

Updated motivation in the OP.

hebasto commented on src/net.h:
```diff
@@ -208,8 +210,18 @@ class CConnman
+    };
+
+    template <typename Callable>
+    void ForEachNode(Callable&& func) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
[Updated](https://github.com/bitcoin/bitcoin/pull/19668#issuecomment-669231930).

ajtowns committed:
sync.h: Make runtime lock checks require compile-time lock checks

ajtowns commented:
Concept ACK.

I'd suggest going further and annotating the `AssertLock(Not)Held` calls with `EXCLUSIVE_LOCKS_REQUIRED(cs)` (or `!cs`) too. This catches existing missing annotations in `validation.cpp:CheckSequenceLocks` and `wallet/scriptpubkeyman.cpp:DescriptorScriptPubKeyMan::AddDescriptKeyWithDb` so seems worthwhile. Patch at https://github.com/ajtowns/bitcoin/commits/202008-assertlockheld

hebasto commented:
@ajtowns 
> I'd suggest going further and annotating the `AssertLock(Not)Held` calls with `EXCLUSIVE_LOCKS_REQUIRED(cs)` (or `!cs`) too. This catches existing missing annotations in `validation.cpp:CheckSequenceLocks` and `wallet/scriptpubkeyman.cpp:DescriptorScriptPubKeyMan::AddDescriptKeyWithDb` so seems worthwhile. Patch at https://github.com/ajtowns/bitcoin/commits/202008-assertlockheld

Thanks! Done.

ajtowns commented:
Might be good to fix up the lock annotations before removing the ASSERT annotation on AssertLockHeldInternal so you can compile successfully at each commit? Otherwise looks good to me.

hebasto commented:
@ajtowns 
> Might be good to fix up the lock annotations before removing the ASSERT annotation on AssertLockHeldInternal so you can compile successfully at each commit? Otherwise looks good to me.

Done, commits are reordered now.

jnewbery commented:
Concept ACK. Perhaps the **Locking/mutex usage notes** section of developer-notes.md should be updated with the current best practice for annotating/asserting locks.

hebasto commented:
@jnewbery 
> Concept ACK. Perhaps the **Locking/mutex usage notes** section of developer-notes.md should be updated with the current best practice for annotating/asserting locks.

You read my mind :)

Do you want me adding a doc commit to this PR?

jnewbery commented:
> Do you want me adding a doc commit to this PR?

I think it's fine to add it to this PR.

vasild commented on src/sync.h:
```diff
@@ -53,8 +53,9 @@ void LeaveCritical();
 template <typename MutexType>
-void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs);
-void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, void* cs);
+void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
+template <typename MutexType>
+void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs);
```
These can also be tagged with `ASSERT_EXCLUSIVE_LOCK`.

```suggestion
void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
template <typename MutexType>
void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(!cs);
```

vasild replied:
I tried the following with `AssertLockHeldInternal() EXCLUSIVE_LOCKS_REQUIRED(cs) ASSERT_EXCLUSIVE_LOCK(cs);`:

```cpp
void f()
{
    // The following line produces a warning due to the EXCLUSIVE_LOCKS_REQUIRED attribute:
    // Calling function 'AssertLockHeldInternal<AnnotatedMixin<std:: __1::recursive_mutex>>' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
    AssertLockHeld(cs_main);

    // The following line produces a warning due to the ASSERT_EXCLUSIVE_LOCK attribute:
    // Acquiring mutex 'cs_main' that is already held
    LOCK(cs_main);
}
```

However if the assert attribute is defined first: `ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs)` then only the second warning is generated (`EXCLUSIVE_LOCKS_REQUIRED` is ignored in this case).

hebasto replied:
> These can also be tagged with `ASSERT_EXCLUSIVE_LOCK`.

What are expected benefits?

hebasto replied:
@vasild 
Here are branches to test your suggestion:
- https://github.com/hebasto/bitcoin/tree/pr16981-base
- https://github.com/hebasto/bitcoin/tree/pr16981-suggestion

Compiling the branch with suggestion miss the following warnings:
```
wallet/scriptpubkeyman.cpp:1861:9: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_desc_man' [-Wthread-safety-analysis]
        m_map_crypted_keys[pubkey.GetID()] = make_pair(pubkey, crypted_secret);
        ^
wallet/scriptpubkeyman.cpp:1864:9: warning: reading variable 'm_map_keys' requires holding mutex 'cs_desc_man' [-Wthread-safety-analysis]
        m_map_keys[pubkey.GetID()] = key;

```

and

```
wallet/wallet.cpp:4429:54: warning: passing variable 'vMasterKey' by reference requires holding mutex 'cs_wallet' [-Wthread-safety-reference]
                if (!spk_manager->CheckDecryptionKey(vMasterKey) && !spk_manager->Encrypt(vMasterKey, nullptr)) {
                                                     ^
wallet/wallet.cpp:4429:91: warning: passing variable 'vMasterKey' by reference requires holding mutex 'cs_wallet' [-Wthread-safety-reference]
                if (!spk_manager->CheckDecryptionKey(vMasterKey) && !spk_manager->Encrypt(vMasterKey, nullptr)) {

```
```

vasild replied:
Ok, I confirm the above. The presence of `ASSERT_EXCLUSIVE_LOCK()` causes some warnings to be silenced.

aaronpuchert replied:
> However if the assert attribute is defined first: `ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs)` then only the second warning is generated (`EXCLUSIVE_LOCKS_REQUIRED` is ignored in this case).

This is somewhat expected. If you assert `cs`, it's assumed to be held afterwards. If you require it to be held and then assert it, we warn if it isn't initially held, and then assume it to be held afterwards.

Generally though you shouldn't expect attributes to be treated in the order they are given, and this particular combination of attributes just doesn't make sense. (Either you're assuming a capability that you've just asserted, which is redundant, or you assert a capability that you've previously assumed, which is pointless.)

vasild replied:
@aaronpuchert,
> This is somewhat expected. If you assert cs, it's assumed to be held afterwards. If you require it to be held and then assert it, we warn if it isn't initially held, and then assume it to be held afterwards.

I agree with that if they were two distinct functions:
```cpp
func_tagged_with_assert_exclusive_lock();
...
func_tagged_with_exclusive_lock_required();
...
```

But for a single function: `void f() ATTR1 ATTR2` I wouldn't expect the behavior to differ if defined like `void f() ATTR2 ATTR1` instead.

If the combination of `ATTR1` and `ATTR2` is meaningless then maybe it would be good to produce some warning or error instead of silently assuming some behavior. Like for `signed unsigned int x;`.

aaronpuchert replied:
> If the combination of `ATTR1` and `ATTR2` is meaningless then maybe it would be good to produce some warning or error instead of silently assuming some behavior.

Makes sense, but there are many attributes and I'm not sure whether we can exhaustively figure out which combinations make sense and which don't.

hebasto committed:
doc: Add best practice for annotating/asserting locks

hebasto commented:
> > Do you want me adding a doc commit to this PR?
> 
> I think it's fine to add it to this PR.

Done.

@jnewbery @jonatack Mind checking wording in the doc commit?

vasild commented on doc/developer-notes.md:
```diff
@@ -746,6 +746,72 @@ the upper cycle, etc.
+{
+public:
+    ...
+    mutable RecursiveMutex cs;
```
nit: Maybe use an example that does not involve `RecursiveMutex` since we recommend against using it and if we manage to obliterate it from the code, then this example will require adjusting.

hebasto replied:
Oh, I've failed to find an example in our code base that embraces a `Mutex` instance, an annotation in function member declaration, and an assertion in function member definition.

vasild reviewed (approved):
ACK 766682366

vasild commented:
This PR basically replaces the attribute `ASSERT_EXCLUSIVE_LOCK()` of `AssertLockHeldInternal()` with `EXCLUSIVE_LOCKS_REQUIRED()` and fixes some of the exposed warnings.

It is unfortunate that we can't have both.

jnewbery commented:
This all seems reasonable and the updated developer docs are clear and make sense.

@ajtowns : what does your :-1: mean here https://github.com/bitcoin/bitcoin/pull/19668#issuecomment-671313972? Are you saying that we can have both attributes, or that it's not unfortunate, or that we can't have both and that this PR is going in the wrong direction? For someone who doesn't know very much about the safety annotations, it's difficult for me to interpret what your objection is.

hebasto commented:
Rebased 76668236610198f61bf60c42a40c2826585515b7 -> 9ee46539e2c65e4ac53149fcdf1badeb35279013 ([pr19668.05](https://github.com/hebasto/bitcoin/commits/pr19668.05) -> [pr19668.06](https://github.com/hebasto/bitcoin/commits/pr19668.06)) due to the conflict with #19704.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -628,13 +628,12 @@ static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
         }
         connman.ForNode(nodeid, [&connman](CNode* pfrom){
-            AssertLockHeld(cs_main);
+            LockAssertion lock(::cs_main);
```
LockAssertion's `__attribute__` is `EXCLUSIVE_LOCK_FUNCTION` aka `ACQUIRE`. This goes against the clang documentation, which states that the capability is *not* held. https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquire-acquire-shared-release-release-shared

> ACQUIRE is an attribute on functions or methods, which declares that the function acquires a capability, but does not release it. The caller must not hold the given capability on entry, and it will hold the capability on exit. ACQUIRE_SHARED is similar.

hebasto replied:
https://github.com/bitcoin/bitcoin/blob/4631dc5c578475fd3ca7a91676f7daf788a11192/src/sync.h#L353-L355

> _when it_ [the fact that mutex is locked] _couldn't be determined otherwise_ [by Thread Safety Analysis]

Such lambda expression is that case.

MarcoFalke replied:
But in that case the `AssertLockHeld` (version as in current master) should be used. That would also comply with the clang docs:  https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability

> These are attributes on a function or method that does a run-time test to see whether the calling thread holds the given capability. The function is assumed to fail (no return) if the capability is not held.

I think that the `AssertLockHeld` is currently overused, as we use it as an run-time backup of the compile-time check. It might be good to add an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings.

MarcoFalke replied:
The alternative version could even be tagged with `EXCLUSIVE_LOCKS_REQUIRED`, so that correct usage of the run-time-only check would be enforced at compile time.

hebasto replied:
This PR goal is decouple compile-time check from run-time one.

hebasto replied:
From _the point of Thread Safety Analysis_ the `AssertLockHeld()` is a run-time check now. That hides compile-time checks.

This PR makes `AssertLockHeld()` a compile-time check (again from _the point of Thread Safety Analysis_).

> It might be good to add an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings.

This PR makes (not adds) "an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings."

MarcoFalke replied:
Ok, but why are you not adding the annotation that the documentation suggests for this use case?

```diff
diff --git a/src/sync.h b/src/sync.h
index 7b397a8003..4a222442a9 100644
--- a/src/sync.h
+++ b/src/sync.h
@@ -357,7 +357,7 @@ public:
 struct SCOPED_LOCKABLE LockAssertion
 {
     template <typename Mutex>
-    explicit LockAssertion(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
+    explicit LockAssertion(Mutex& mutex) ASSERT_EXCLUSIVE_LOCK(mutex)
     {
 #ifdef DEBUG_LOCKORDER
         AssertLockHeld(mutex);
```
 https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability

>    These are attributes on a function or method that does a run-time test to see whether the calling thread holds the given capability. The function is assumed to fail (no return) if the capability is not held.

hebasto replied:
> Ok, but why are you not adding the annotation that the documentation suggests for this use case?

It seems a bit out of this PR scope.
Nevertheless, `ASSERT_EXCLUSIVE_LOCK` "does a run-time test". But `LockAssertion` provides compile-time TSA check. See: https://github.com/bitcoin/bitcoin/pull/14437/files#diff-df95fc819b237793cb462142a31b1871

MarcoFalke replied:
The run time check was added when it was renamed from "annotation" to "assertion": https://github.com/bitcoin/bitcoin/pull/16034#issuecomment-493042117

hebasto replied:
Our crafted run time check `AssertLockHeld()` is preferred over TSA run time check `ASSERT_EXCLUSIVE_LOCK()`, no?

ajtowns replied:
I think the only actual differences between `EXCLUSIVE_LOCK_FUNCTION` and `ASSERT_EXCLUSIVE_LOCK` are that the latter doesn't warn you when it sees you attempt to lock the same thing twice (so whether a redundant assertion like `LOCK(x); LockAssertion a(x);` gives an error or not) and whether it warns you if you don't release the lock when you leave the scope -- but since we've got an unlocking destructor here I don't think that matters much. [Ref](https://github.com/llvm/llvm-project/blob/a5bb24758d02e1e42b72561d0ddbd682776d1a4d/clang/lib/Analysis/ThreadSafety.cpp#L1833-L1835)

So I think it's fine to mark this function with either `EXCLUSIVE_LOCK_FUNCTION` or `ASSERT_EXCLUSIVE_LOCK`. `E_L_F` has the advantage that it discourages using `LockAssertion` where it's not necessary; otoh `A_E_L` has the advantage that it won't cause problems if someday clang becomes smarter about when locks are held, and we need the assertion for old clang versions but not new ones. Given there's not many docs about how the `ASSERT` variant works and the example code doesn't actually exercise it, probably safer to stick with `EXCLUSIVE_LOCK_FUNCTION`?

MarcoFalke replied:
Sure. Lets leave it as is. Resolving discussion.

aaronpuchert replied:
`ASSERT_EXCLUSIVE_LOCK` was created for exactly this purpose, and it's conceptually the right attribute to use here. `EXCLUSIVE_LOCK_FUNCTION` isn't right for the very simple reason that you don't actually lock. Also this shouldn't be a `SCOPED_LOCKABLE` because assertions don't go out of scope.

> `E_L_F` has the advantage that it discourages using `LockAssertion` where it's not necessary

It's redundant but obviously safe, so it's deliberate that we don't warn about that.

> Given there's not many docs about how the `ASSERT` variant works

Feel free to suggest changes. We obviously want to keep the docs short so that you don't have to read a novel, but they could be a bit more verbose.

> and the example code doesn't actually exercise it,

You can have a look at [the test](https://github.com/llvm/llvm-project/blob/a5bb24758d02e1e42b72561d0ddbd682776d1a4d/clang/test/SemaCXX/warn-thread-safety-analysis.cpp#L4433-L4531).

maflcko reviewed (commented):
It seems to be doing the exact opposite of what the clang docs say, so it would be good to explain why this needs to be different

MarcoFalke commented on src/sync.h:
```diff
@@ -53,7 +53,7 @@ void LeaveCritical();
 std::string LocksHeld();
 template <typename MutexType>
-void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs);
+void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs);
```
in commit 5971934eac13809143c763aa515b1960a7368d74:

why are you removing this? This should come with the appropriate annotation to enforce correct usage and warn about missing compile-time annotations

```suggestion
void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
```

MarcoFalke replied:
In fact, adding this annotation and compiling the commit shows one instance of a missing annotation.

MarcoFalke replied:
Ah, it looks like you already did that in the next commit :man_facepalming:  653c8d065d90642ffb2d8d4d49474b4abc67a811

hebasto replied:
Should I reorder/squash commits?

MarcoFalke replied:
No need to reorder, but if you want you can describe the two preparatory commits. Suggested git range-diff for the first two commits:


```
1:  00e537d140 ! 1:  efeaeaec78 Use LockAssertion utility class instead of AssertLockHeld()
    @@ Metadata
      ## Commit message ##
         Use LockAssertion utility class instead of AssertLockHeld()
     
    +    First, add the proper attribute to LockAssertion, according to
    +    https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability
    +
    +    Then, to prepare for commit "Do not hide compile-time thread safety
    +    warnings", replace AssertLockHeld with LockAssertion where needed.
    +
      ## src/net_processing.cpp ##
     @@ src/net_processing.cpp: static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
                  }
    @@ src/net_processing.cpp: void PeerLogicValidation::EvictExtraOutboundPeers(int64_
      
                      // Only disconnect a peer that has been connected to us for
                      // some reasonable fraction of our check-frequency, to give
    +
    + ## src/sync.h ##
    +@@ src/sync.h: public:
    + struct SCOPED_LOCKABLE LockAssertion
    + {
    +     template <typename Mutex>
    +-    explicit LockAssertion(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
    ++    explicit LockAssertion(Mutex& mutex) ASSERT_EXCLUSIVE_LOCK(mutex)
    +     {
    + #ifdef DEBUG_LOCKORDER
    +         AssertLockHeld(mutex);
2:  0ed94d8f86 ! 2:  f0e821b7ef Add missed thread safety annotations
    @@ Metadata
      ## Commit message ##
         Add missed thread safety annotations
     
    +    This is needed for commit "sync.h: make runtime lock checks require
    +    compile-time lock checks" to pass
    +
      ## src/wallet/scriptpubkeyman.h ##
     @@ src/wallet/scriptpubkeyman.h: private:
          //! keeps track of whether Unlock has run a thorough check before

hebasto replied:
Thanks! Commit messages have been updated.

MarcoFalke commented:
Just one question: https://github.com/bitcoin/bitcoin/pull/19668#discussion_r479669946

Otherwise, almost ACK 984ea63cb6 🔘

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

almost ACK 984ea63cb6 🔘
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgRiQv+OzT2oaGMTBdl/bL/KCX2vXdlLOYUopemBZbi3N+ZZxDTGcYwPBWDsjoF
1BAKpoRQfy18UvI+TwftS4hccqt2676r332uahJ3TD0SwT5LGe4IyR6w2QAnB86o
Wv7eLRJIMoAb1b3UfL2wTNCEWXouxqoAVeEsucq4PyfI5lDQKpilJSSvJjRmvScK
+wuNfvhmXeMX+RzYBTBilG0SELfjhnsNSn9l+rkmMgbD7kAzHatUFJ+FeSRHDJPb
Lwr0XBI6hsYYCn/d7vH9o2PVXcULJ4PFpjg5HU4/f40dF+7vmva38CA9cKlJyVEU
2McPNT8h99/JAonUFMTFAzroAjAKyD1SVDRl2vyKAArT84HGMnKH0cr4e5WKaHuw
oa4DXt0fyz9PxZayq9hd3SPOJB4ukxrY8rWI2QFZHgOU5AUAKR0L8B1fJA+KKFgU
zVGxPz38xUO2UahyHqzN6Nm/ELyXgBq9lZG0xM5kFzL1Xha/dsooqvu0M8faSzSM
+0SGxDw8
=jU5f
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `7b19e59265341027936672d6d3e2e75fba6bee9ec10e1cd33960b42ec4416ed8  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401087b19e59265341027936672d6d3e2e75fba6bee9ec10e1cd33960b42ec4416ed8f0106ea759232aa89ded4ec0763b06c49eec08fff01040de79ca7d0e6b2acb9b3ff4feff3ce408f1045f4a90eaf008ec84146f72e044db0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01041ea7556861d4239bccad0e609fe172608f1045f4a90e9f008c3362913b872631a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d8f54fa7b4dde9459b331c80d5763e6708f1045f4a90e9f00806e4dba08b3ceb1c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010da21bd7025188b1963af63ae1e87b8b208f1045f4a90e9f0083f088945febd9b940083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

hebasto commented:
Updated 9ee46539e2c65e4ac53149fcdf1badeb35279013 -> ea74e10acf17903e44c85e3678853414653dd4e1 ([pr19668.06](https://github.com/hebasto/bitcoin/commits/pr19668.06) -> [pr19668.07](https://github.com/hebasto/bitcoin/commits/pr19668.07), [diff](https://github.com/hebasto/bitcoin/compare/pr19668.06..pr19668.07)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19668#discussion_r479671611):
> No need to reorder, but if you want you can describe the two preparatory commits.

MarcoFalke commented:
ACK ea74e10acf 🎙

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK ea74e10acf 🎙
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjduAwAvJQy8WxET79bR7T5ZL/UoKzl6VDAiNEM5pbrfO0xCRrsD1bM+s+RABtR
nxkeaGOrOrUNAvbGun0GKtT7voqp3x3RwXtcUdeNRCSJXp5zOIzMP8V87FCzMZ9v
WI9jJh2E6JngRXTzJayY7R4enioFcCoLK415AwlEZaa06jycBP3TpAPOsjUczaIC
8EjDku/22miUzPcwnHAw1XR9bhXZ3oyl6b7Lwx4mktiGFCNtvr54MLopFls//+bs
ZxFQ3Wqma5wLm8aBPCfHkP1ZkTtLzEZnH9ho2fphW986fDjvd+G0+H00/zV1P8ey
QeY+fTU0zzxoUMXUQsoGvXAxcfRnV1/EtzWtqraTYs+ebQx1kN8PPilkBcIioyI/
llP9i7SP/QJEH74bxQPD6IaC8NlyY9is2RI99DzBiGYawB0zZj65ri9hWIocD+XD
LYefVtHoAjaEflzHGnYWNiEGBzSOEm/RUUONBT1JBFmkvFQC6LNWoYGbZe+MfuYT
/at0I08D
=0zQn
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `45b97e7100d59a8015d51f47d4403a42ca4a6457ae8da97a2cbcf721bf8f5f59  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010845b97e7100d59a8015d51f47d4403a42ca4a6457ae8da97a2cbcf721bf8f5f59f0100e191ac28fbc019bc49f7dfe3789bc7608fff0102c889f92f9368ceedc7a23abacce448108f1045f4a9739f00873dce31db7546a340083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01058ff1688ef2bea2330846ca69059dd2308f1045f4a9738f008588df02151893f080083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ae1785b9b7183a1d0ecf5d6661309b0c08f1045f4a9739f008f15dacab759f2e110083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d7f01a051af86d017c519d14a91a7f1308f1045f4a9739f008f4f7d567a4448c590083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

hebasto commented:
@ajtowns @jnewbery @vasild

Begging for another review round :)

jnewbery commented:
ACK ea74e10acf17903e44c85e3678853414653dd4e1

ajtowns commented:
ACK ea74e10acf17903e44c85e3678853414653dd4e1

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -1327,7 +1326,7 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
 
     m_connman.ForEachNode([this, &pcmpctblock, pindex, &msgMaker, fWitnessEnabled, &hashBlock](CNode* pnode) {
-        AssertLockHeld(cs_main);
+        LockAssertion lock(::cs_main);
```
What's the reason to prefer LockAssertion to AssertLockHeld if AssertLockHeld works at compile time and runtime and LockAssertion only works at compile time? Maybe this could be clarified https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization

hebasto replied:
When TSA fails to detect that a mutex is locked we do want to use `LockAssertion` as using `AssertLockHeld` no longer helps:
```
$ git diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 0e049bd66..90bd626fc 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -1370,7 +1370,7 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     }
 
     m_connman.ForEachNode([this, &pcmpctblock, pindex, &msgMaker, fWitnessEnabled, &hashBlock](CNode* pnode) {
-        LockAssertion lock(::cs_main);
+        AssertLockHeld(::cs_main);
 
         // TODO: Avoid the repeated-serialization here
         if (pnode->nVersion < INVALID_CB_NO_BAN_VERSION || pnode->fDisconnect)
$ make
net_processing.cpp:1373:9: warning: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::recursive_mutex> >' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        AssertLockHeld(::cs_main);
        ^
./sync.h:79:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
net_processing.cpp:1378:9: warning: calling function 'ProcessBlockAvailability' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        ProcessBlockAvailability(pnode->GetId());
        ^
net_processing.cpp:1379:30: warning: calling function 'State' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
        CNodeState &state = *State(pnode->GetId());
                             ^
net_processing.cpp:1383:18: warning: calling function 'PeerHasHeader' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
                 ^
net_processing.cpp:1383:51: warning: calling function 'PeerHasHeader' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
                !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) {
                                                  ^
5 warnings generated.
```

hebasto replied:
And this `LockAssertion` application is documented: https://github.com/bitcoin/bitcoin/blob/a0a422c34cfd6514d0cc445bd784d3ee1a2d1749/src/sync.h#L355-L357

hebasto replied:
> Maybe this could be clarified https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#threads-and-synchronization

Mind suggesting the clarification as my English is so poor?

ryanofsky replied:
Oh, I wasn't aware `AssertLockHeld` annotations changed from `ASSERT_EXCLUSIVE_LOCK` 
to `EXCLUSIVE_LOCKS_REQUIRED` in #19668. This seems like a questionable decision to me, because now instead of being able to use `AssertLockHeld` everywhere, we have separate confusing `AssertLockHeld` and `LockAssertion` calls.

I think I would want to clarify this by changing `AssertLockHeld` annotations back to `ASSERT_EXCLUSIVE_LOCK`, making it a weaker assertion that only causes errors at runtime, never compile time. Then we could drop LockAssertion and s/LockAssertion/AssertLockHeld/ and only have one type of assertion to think about.

If this doesn't work because it interferes with work you are doing, another option 
might be to s/LockAssertion/WeaklyAssertLockHeld/ everywhere. (Also maybe simplify the implementation to be an `ASSERT_EXCLUSIVE_LOCK` function instead of a class instance that requires a variable declaration.) The new naming should make it clearer which call is preferred:

* `AssertLockHeld` - stricter assertion which causes errors at compile time if the compiler supports annotations, or causes errors at runtime if the compiler doesn't support annotations
* `WeaklyAssertLockHeld` - less strict alternative which only triggers errors at runtime, never compile-time, and can be used when it's not possible for compiler to determine that the mutex is locked.

ajtowns replied:
We have three types of assertions related for locking:

* marking a function as `EXCLUSIVE_LOCKS_REQUIRED` has a compile time check with clang that the caller has already obtained the lock
* declaring a `LockAssertion` instance overrides the clang compile time checks by saying that we've already somehow acquired the lock in a way that we can't prove via clang thread safety annotations. This is useful for lambda functions (which could be annotated) that are called from some generic dispatcher like `ForEach` (that can't be annotated).
* adding a call to `AssertLockHeld` does a runtime check if DEBUG_LOCKORDER is enabled. it's annotated with EXCLUSIVE_LOCKS_REQUIRED so is redundant with the compile time checks, but the compile time checks aren't available outside of clang. it's automatically called from LockAssertion to ensure that the compile time checks aren't overridden incorrectly.

As a result:

* using AssertLockHeld everywhere remains fine, and no longer needlessly disables the compile time checks
* LockAssertion should only be used very rarely -- and in fact it's only used in net_processing in some ForEach/ForEachNode functions.

Maybe we should move the locking outside of the ForEach functions into the caller, and convert them into `for ( x : nodes ) { ... }` instead, in which case we could probably do away with LockAssertion entirely.

ryanofsky replied:
> no longer needlessly disables the compile time checks

It's funny the only reason I ever called it was to disable compile time checks, and now it doesn't do that anymore! :smile: If really the only thing AssertLockHeld is doing is adding runtime checks on top of compile time checks that are already present, I think that is nearly useless. Anyway, can take it up in another PR.

ryanofsky replied:
> Anyway, can take it up in another PR.

Created #19865 to experiment

ajtowns replied:
You use `LockAssertion g(cs_foo);` to tell the compiler that a lock is guaranteed to already be held within a code block even though the function annotations don't guarantee it. AssertLockHeld is useful if you're not compiling with clang.

ryanofsky replied:
> AssertLockHeld is useful if you're not compiling with clang.

Yes, that's what I mean by "nearly useless." It can help developers who aren't using clang, and who are compiling with DEBUG_LOCKORDER, and who want to wait for an assert to trigger at runtime. But it won't make the codebase more thread safe because we already don't merge changes that don't compile with clang.

Any case, if you see problems with #19865, it would be good to comment there. I'm not an expert on lock annotations, so maybe the AssertLockHeld calls removed in #19865 are actually doing something useful besides haphazardly repeating compile time assertions at runtime.

hebasto replied:
> > AssertLockHeld is useful if you're not compiling with clang.
> 
> Yes, that's what I mean by "nearly useless." It can help developers who aren't using clang, and who are compiling with DEBUG_LOCKORDER, and who want to wait for an assert to trigger at runtime. But it won't make the codebase more thread safe because we already don't merge changes that don't compile with clang.

I [disagree](https://github.com/bitcoin/bitcoin/pull/19865#issuecomment-686986462) with statements about "useless" `AssertLockHeld`.
